R/derive_param_framingham.R

```diff
@@ -0,0 +1,370 @@
+#' HDL serum cholesterol (mg/dL), sex, smoking status, diabetic status,
+#' and treated for hypertension flag.
+#'
+#' @param sysbp Systolic blood pressure
+#'
+#'   A numeric vector is expected.
+#'
+#' @param chol Total serum cholesterol (mg/dL)
+#'
+#'   A numeric vector is expected.
+#'
+#' @param cholhdl HDL serum cholesterol (mg/dL)
+#'
+#'   A numeric vector is expected.
+#'
+#' @param age Age (years)
+#'
+#'   A numeric vector is expected.
+#'
+#' @param sex Gender
+#'
+#'   A character vector is expected.
+#'
+#' @param smokefl Smoking Status
+#'
+#'   A character vector is expected.
+#'
+#' @param diabetfl Diabetic Status
+#'
+#'   A character vector is expected.
+#'
+#' @param trthypfl Treated for hypertension status
+#'
+#'   A character vector is expected.
+#'
+#' @author Alice Ehmann
+#'
+#' @details
+#' The predicted probability of having cardiovascular disease (CVD)
+#' within 10-years according to Framingham formula D'Agostino, 2008 is:
+#'
+#' For women:
+#'   Age Factor = 2.32888;
+#'   Total Chol Factor = 1.20904;
+#'   HDL Chol Factor = -0.70833;
+#'   SysBPFactor=2.76157 if not on hypertension medications, 2.82263 if on
+#'     hypertension medications, Avg Risk = 26.1931 and Risk Period;
+#'   Smoker=.52873 if smoker, 0 otherwise;
+#'   Diabetes Present=.69154 is present, 0 otherwise;
+#'   Avg Risk = 26.1931;
+#'   RiskPeriodFactor = 0.95012;
+#'
+#' For men:
+#'   Age Factor = 3.06117;
+#'   Total Chol Factor = 1.12370;
+#'   HDL Chol Factor = -0.93263;
+#'   SysBPFactor=1.93303 if not on hypertension medications, 2.99881 if on
+#'     hypertensino medications,;
+#'   Smoker=.65451 if smoker, 0 otherwise;
+#'   Diabetes Present=.57367 is present, 0 otherwise,;
+#'   Avg Risk = 23.9802;
+#'   Risk Period Factor = 0.88936;
+#'
+#' Equation:
+#' \deqn{RiskFactors = (log(Age) * AgeFactor)
+#' + (log(TotalChol) * TotalCholFactor)
+#' + (log(CholHDL) * CholHDLFactor)
+#' + (log(SysBP) * SysBPFactor) + Smoker
+#' + Diabetes Present - AvgRisk}
+#'
+#' \deqn{Risk = 100 * (1 - RiskPeriodFactor ^ exp(RiskFactors))}
+#'
+#' @return A numeric vector of Framingham values
+#'
+#' @keywords computation
+#'
+#' @export
+#'
+#' @examples
+#' compute_framingham(sysbp = 133, chol = 216.16, cholhdl = 54.91, age=53,
+#'                    sex = "M", smokefl = "N", diabetfl = "N", trthypfl = "N")
+#'
+#' compute_framingham(sysbp = 161, chol = 186.39, cholhdl = 64.19, age=52,
+#'                    sex = "F", smokefl = "Y", diabetfl = "N", trthypfl = "Y")
+#'
+compute_framingham <- function(sysbp, chol, cholhdl, age, sex, smokefl,
+                               diabetfl, trthypfl) {
+
+  assert_numeric_vector(sysbp)
+  assert_numeric_vector(chol)
+  assert_numeric_vector(cholhdl)
+  assert_numeric_vector(age)
+  assert_character_vector(sex, values = c("M", "F"))
+  assert_character_vector(smokefl, values = c("Y", "N"))
+  assert_character_vector(diabetfl, values = c("Y", "N"))
+  assert_character_vector(trthypfl, values = c("Y", "N"))
```

Comment:
Should we accept `NA_character_`? Otherwise, the user would need to exclude observations with missing information in a preprocessing and re-add them after the `compute_framingham()` call, e.g., when deriving a variable in ADSL.

For `SEX` the values `"UNDIFFERENTIATED"` and `"U"` are CDISC standard values. Should we consider them?